### PR TITLE
Scala 3 cross compile

### PR DESCRIPTION
**Pull request by fedefernandez**
### Config changes :gear:

* Includes a minimal `.gitignore` file
* Adds a `build.properties` file
* Adds cross compilation to Scala `2.13.x` and `3.2.x`. Scala `2.13.x` by default.
* :information_source: Updates `build.sbt` to use `CompositeProject`. This allows:
  * Specify modules names and other settings (like dependencies)
  * Remove the modules `build.sbt` files
* Updates to sbt 1.x slash syntax (`Compile / scalaSource` vs `scalaSource in Compile`)
* Bumps `scalatest` to `"3.2.15"` and sets the scope to `Test`

### Code changes that need attention 👀 
* [Lazy Evaluation/Lazy Lists/test/Test.scala](https://github.com/jetbrains-academy/scala-tutorial/compare/master...47degrees:scala-tutorial:scala-3?expand=1#diff-674bb666f8b9f620172b3032e8ca905638e5989b45064d2ce9dade4c44b36458)
As far as I can see, the `LazyList` implementation for Scala 2.13.x and Scala 3 seems more performant and do one less iteration. The test has been updated appropriately. 
* [Polymorphic Types/Covariance/src/Covariance.scala](https://github.com/jetbrains-academy/scala-tutorial/compare/master...47degrees:scala-tutorial:scala-3?expand=1#diff-163868f82e5def293db3114a9ee43d2ac87b7820af18aae7c8d8401259b896ea)
Scala 3 fail to compile due to the missing type
```bash
[error] -- [E171] Type Error: Polymorphic Types/Covariance/src/Covariance.scala:7:23
[error] 7 |  class Mammal extends Animal
[error]   |                       ^^^^^^
[error]   |missing argument for parameter animal of constructor Animal in class Animal: (animal: T): Covariance.Animal[T]
```
Also, the exercise presents an extra layer of complexity not needed for explaining the concept (a wrapper type depending on itself). I've changed the supertype to `AnimalKind` to make it independent from the generic one (`Animal[T]`) 

### Other aesthetic code changes 💅🏼 
* Changes the scalatest import for the new version (`import org.scalatest.matchers.should.Matchers`)
* Adds whitespace before brackets
* Removes `extends FlatSpec with Matchers` from main files (not used)

### Automatic changes by the EduTools plugin :robot:
* Added a break line on every `.yaml` file
* Replaced `EduTools` with `JetBrains Academy plugin` 🤷🏼‍♂️ 

## Next Steps

* Course preview review. I've reviewed the course by myself, but a second pair of 👀 is welcome
* As far as I know, EduTools for Scala doesn't support cross-building, so students will see a Scala 2 project (the default in this PR). To generate a Scala 3 course, the setting `scalaVersion` needs to be changed before creating the archive.

### CODE REVIEW

File: Lazy Evaluation/Lazy Lists/test/Test.scala [hunk 1]
```diff
@@ -2,13 +2,13 @@ import LazyEvaluation.{llRange, rec}
...
   def `test encoding`(): Unit = {
     llRange(1, 10).take(3).toList
-    rec shouldBe 4
+    rec shouldBe 3
     llRange(1, 10).take(1).toList
-    rec shouldBe 6
+    rec shouldBe 4
     llRange(1, 10).take(2).toList
-    rec shouldBe 9
+    rec shouldBe 6
```
Comment (julienrf) [inline Lazy Evaluation/Lazy Lists/test/Test.scala:12]: The assertions look good to me, but the fact that we were observing a different behavior with scala-collection-compat means that there is probably a bug in scala-collection-compat… :thinking:

File: Structuring Information/Introduction/task-info.yaml [hunk 2]
```diff
@@ -5,12 +5,10 @@ files:
 - name: src/StructuringInformation.scala
   visible: true
   placeholders:
-  - offset: 129
+  - offset: 130
     length: 23
-    placeholder_text: //the function should return a new note with name C, duration
-      of quarter and a third octave
+    placeholder_text: "//the function should return a new note with name C, duration\
+      \ of quarter and a third octave"
```
Comment (julienrf) [inline Structuring Information/Introduction/task-info.yaml:11]: I am not sure about the formatting here, would the CI fail if there was an error?

File: Polymorphic Types/Covariance/src/Covariance.scala [hunk 3]
```diff
@@ -1,9 +1,10 @@
 object Covariance {
   class Animal[+T](val animal: T)
 
-  class Reptile extends Animal
+  trait AnimalKind
```
Comment (julienrf) [inline Polymorphic Types/Covariance/src/Covariance.scala]: I am really surprised that this code compiles with Scala 2… Thank you for fixing it!

However, to be honest the code in this file does not make a lot of sense to me, and I don’t see it in the page https://www.scala-exercises.org/scala_tutorial/polymorphic_types. Is it actually used somewhere?

File: Structuring Information/Introduction/task-info.yaml (hunk 2 above)
Comment (julienrf) [reply Structuring Information/Introduction/task-info.yaml:11]: Oh, I see that there is no CI… Could you please check and confirm that this works?

File: Lazy Evaluation/Lazy Lists/test/Test.scala (hunk 1 above)
Comment (fedefernandez) [reply Lazy Evaluation/Lazy Lists/test/Test.scala:12]: 👍🏼 Let me dig into that

File: Structuring Information/Introduction/task-info.yaml (hunk 2 above)
Comment (fedefernandez) [reply Structuring Information/Introduction/task-info.yaml:11]: I can confirm this works, as the tests pass locally, and the course preview is properly generated

File: Polymorphic Types/Covariance/src/Covariance.scala (hunk 3 above)
Comment (fedefernandez) [reply Polymorphic Types/Covariance/src/Covariance.scala]: I always try to minimize the changes, but in this case, this exercise deserves some thought and bigger improvement. Let me try something.

File: Structuring Information/Introduction/task-info.yaml (hunk 2 above)
Comment (fedefernandez) [reply Structuring Information/Introduction/task-info.yaml:11]: Screenshot 2023-04-12 at 09 46 33

File: Lazy Evaluation/Lazy Lists/test/Test.scala (hunk 1 above)
Comment (fedefernandez) [reply Lazy Evaluation/Lazy Lists/test/Test.scala:12]: @julienrf you were right in the sense there was a bug in scala-collection-compat, fixed on `2.2.0`

As you can see here, the current version works fine.

https://gist.github.com/fedefernandez/c60824b4af90fa51243c6b0e19befd35

The `LazyList` was introduced in version `2.1.5`. Both, `2.1.5` and `2.1.6` have that bug:

```
4 shouldBe 3
6 shouldBe 4
9 shouldBe 6
```

The fix was included in version `2.2.0`: https://github.com/scala/scala-collection-compat/releases/tag/v2.2.0

Screenshot 2023-04-12 at 10 13 02

Comment (julienrf) [reply Lazy Evaluation/Lazy Lists/test/Test.scala:12]: Thank you very much for the investigation and detailed report! I am glad to see that the bug has been fixed in scala-collection-compat in the meantime :)

**Comment by fedefernandez** [comment]
@julienrf I've updated the course. I keep it simple, but I think this gives the idea to the student. Feedback is welcome. Thanks!

**Review by julienrf — APPROVED** [review]
That looks great to me, thank you!

**Review by Geravant — APPROVED** [review]

**Comment by fedefernandez** [comment]
Thanks, @Geravant.

**Pull request merged**